Add empty, resize, dup

eriknw opened the pull request:
The if-elif-else structure in Cython is pretty annoying, but ¯\\_(ツ)_/¯

Also, these required adding to the classes in C++.  Hopefully this is okay ¯\\_(ツ)_/¯

@jim22k, does this satisfy your requests?

eriknw committed:
Add functions to perform empty, dup, and resize.

eriknw committed:
Merge branch 'main' into empty_resize_dup

eriknw committed:
black

jim22k reviewed (changes requested):
I sent you an email with details. Essentially, I need an overhaul of `resize` because sparsity makes things more complicated than one might like.

eriknw committed:
Split resize into `resize_{pointers,index,values}`

eriknw committed:
oops pointers

eriknw commented:
Updated.

eriknw committed:
Add `dup`, `empty_like`, `empty`, `resize_dim`

Also, switched to use `void*` for things.
Haven't yet deleted the old `empty_*` functions.

eriknw commented:
New functions added.  I need to add tests for them.

Jim, if they work for you, then we can probably remove `empty_mlir_sparse_tensor_fast` and `empty_mlir_sparse_tensor_safe`.

eriknw committed:
Add `MLIRSparseTensor.from_raw_pointer`.  UNTESTED!

eriknw committed:
Test `from_raw_pointer`.

This test is pretty basic, and required me to be able to set `.data` to 0 so it won't be freed twice.

eriknw committed:
Fix (and test) `empty_like`

jim22k committed:
Allow !llvm.ptr<i8> to be returned from MLIR functions

jim22k committed:
Merge branch 'main' into HEAD

jim22k committed:
Formatting
